RecoverCRPVM/Common-Functions.psm1

```diff
@@ -0,0 +1,408 @@
+        return $false
+    }
+     
+    if ($vm.StorageProfile.OsDisk.ManagedDisk)
+    {
+        Write-log "VM ==> $($vm.Name) is a Managed VM, and is currently not supported by this script, cannot continue exiting." -color Red
+        Return $false
+    }
+
+    #Checks to see if the Image exist, if not it returns false as disk swap does not works unless the imgage is available.
+    Try
+    {
+        if ($vm.StorageProfile.OsDisk.CreateOption -eq "FromImage")
+        {
+            $ImageObj =(get-azurermvmimage -Location $vm.Location -PublisherName $vm.StorageProfile.ImageReference.Publisher -Offer $vm.StorageProfile.ImageReference.Offer -Skus $vm.StorageProfile.ImageReference.sku)[-1]
+            if (-not $ImageObj)
+            {
+                Write-Log "Artifact: VMImage was not found,script can't be used you may hit the same error if you manually try to perform the same steps as well." -color red
+                return $false
+            }
+        }
+    }
+    catch
+    {
+        Write-Log "Artifact: VMImage was not found,script can't be used you may hit the same error if you manually try to perform the same steps as well." -color red
+        return $false
+    }
+
+    <#if ($vm.StorageProfile.OsDisk.OsType -ne "Windows")
+    {
+        Write-log "VM ==> $($vm.Name) is not a Windows VM, and is currently not supported by this script, cannot continue exiting." -color Red
+        return $false
+    } #>   
+    return $true
+}
+
+function CreateRescueVM(
+    [Object[]]$vm,
+    [Parameter(mandatory=$true)]
+    [String]$ResourceGroup,
+    [Parameter(mandatory=$true)]
+    [String]$rescueVMNname,
+    [Parameter(mandatory=$true)]
+    [String]$RescueResourceGroup,
+    [String]$prefix = "rescue",
+    [Parameter(mandatory=$false)]
+    [String]$Sku,
+    [Parameter(mandatory=$false)]
+    [String]$Offer,
+    [Parameter(mandatory=$false)]
+    [String]$Publisher,
+    [Parameter(mandatory=$false)]
+    [String]$Version
+    )
+{
+
+    Try
+    {
+        write-log "Initiating the process to create the new Rescue VM" -color Yellow
+        
+        if ($vm.StorageProfile.OsDisk.ManagedDisk) {$managedVM = $true} else  {$managedVM = $false}
+
+        $osDiskName  = $vm.StorageProfile.OsDisk.Name
+        $vmSize = $vm.HardwareProfile.VmSize
+        $osType = $vm.StorageProfile.OsDisk.OsType
+        $location = $vm.Location
+        $networkInterfaceName = $vm.NetworkProfile.NetworkInterfaces[0].Id.split('/')[-1]
+        $MaxStaorageAccountNameLength=24
+
+        $rescueOSDiskName = "$prefix$osDiskName"
+        if (-not $managedVM)
+        {
+            $osDiskVhdUri = $vm.StorageProfile.OsDisk.Vhd.Uri
+            $storageAccountName = $vm.StorageProfile.OsDisk.Vhd.Uri.Split('//')[2].Split('.')[0]
+            $rescueosDiskVhduri = $osDiskVhdUri.Replace($osDiskName,$rescueOSDiskName)
+        }
+
+        $rescuevm = New-AzureRmVMConfig -VMName $rescueVMNname -VMSize $vmSize;
```

Comment:
Please take it as a generic comment, before dereferencing any variable, look for null and empty so that debugging will be easy in future.